extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,337 @@
+            private CompletableFuture<ByteBuffer> future;
+
+            private FragmentState(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+                this.fragmentIndex = fragmentIndex;
+                this.future = future;
+            }
+        }
+
+        /**
+         * Used to cache recently fetched fragments for faster lookup
+         */
+        private final FragmentState[] bufferCache;
+
+        /**
+         * The size of the object in bytes, stored in context to avoid fetching multiple times
+         */
+        private long size;
+
+        S3ChannelContext(final int maxCacheSize) {
+            bufferCache = new FragmentState[maxCacheSize];
+            size = UNINITIALIZED_SIZE;
+        }
+
+        private int getIndex(final int fragmentIndex) {
+            return fragmentIndex % bufferCache.length;
+        }
+
+        void setFragmentState(final int fragmentIndex, final CompletableFuture<ByteBuffer> future) {
+            final int cacheIdx = getIndex(fragmentIndex);
+            final FragmentState cachedEntry = bufferCache[cacheIdx];
+            if (cachedEntry == null) {
+                bufferCache[cacheIdx] = new FragmentState(fragmentIndex, future);
+            } else {
+                // We should not cache an already cached fragment
+                Assert.neq(cachedEntry.fragmentIndex, "cachedEntry.fragmentIndex", fragmentIndex, "fragmentIdx");
+
+                // Cancel any outstanding requests for this cached fragment
+                cachedEntry.future.cancel(true);
+
+                // Reuse the existing entry
+                cachedEntry.fragmentIndex = fragmentIndex;
+                cachedEntry.future = future;
+            }
+        }
+
+        /**
+         * Will return the {@link CompletableFuture} corresponding to provided fragment index if present in the cache,
+         * else will return {@code null}
+         */
+        @Nullable
+        CompletableFuture<ByteBuffer> getCachedFuture(final int fragmentIndex) {
+            final FragmentState cachedFragment = bufferCache[getIndex(fragmentIndex)];
+            if (cachedFragment != null && cachedFragment.fragmentIndex == fragmentIndex) {
+                return cachedFragment.future;
+            }
+            return null;
+        }
+
+        long getSize() {
+            return size;
+        }
+
+        void setSize(final long size) {
+            this.size = size;
+        }
+    }
+
+    private final S3AsyncClient s3AsyncClient;
+    private final String bucket;
+    private final String key;
+    private final S3Instructions s3Instructions;
+
+    /**
+     * The size of the object in bytes, fetched at the time of first read
+     */
+    private long size;
+    private int numFragmentsInObject;
+
+    private SeekableChannelContext channelContext;
+
+    private long position;
+
+    S3SeekableByteChannel(@NotNull final SeekableChannelContext channelContext, @NotNull final URI uri,
```

Comment:
I'm suspicious that the constructor need not take `channelContext`. I assume we'll always immediately call `setContext` before use.